### ISSUE TITLE
Segmentation fault with TorchScript while scripting a function with an argument Optional[List]

### ISSUE THREAD

**Issue by hal-314** [pytorch/pytorch#42859]
## 🐛 Bug

This bug makes "Segmentation fault: 11" in a MacOS Mojave, Python 3.7 (anaconda) and Pytorch 1.5.1 or 1.6. 

## To Reproduce

Run this code have a segmentation fault
```
from typing import Optional, List
import torch
@torch.jit.script
def buggy_fun(x:torch.Tensor, classes:Optional[List]=None):
    
    # Filter by class
    if classes:
        x = x[(x[:, 5:6] == torch.tensor(classes, device=x.device)).any(1)]
```
This code works well if you change `classes:Optional[List]=None` to `classes:Optional[torch.Tensor]=None`.

## Expected behavior

It works or gives an error message but not to kill the python interpreter without any trace.

## Environment
PyTorch version: 1.6.0 or 1.5.1
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.14.6
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.16.4
[conda] _pytorch_select           0.1                       cpu_0  
[conda] _tflow_select             2.3.0                       mkl  
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.4                      233  
[conda] mkl-service               2.0.2            py37h1de35cc_0  
[conda] mkl_fft                   1.0.14           py37h5e564d8_0  
[conda] mkl_random                1.0.2            py37h27c97d8_0  
[conda] msgpack-numpy             0.4.3.2                  py37_0  
[conda] numpy                     1.18.1           py37h7241aed_0  
[conda] numpy-base                1.18.1           py37h6575580_1  
[conda] numpydoc                  0.9.2                      py_0  
[conda] pytorch                   1.6.0                   py3.7_0    pytorch
[conda] pytorch-lightning         0.8.5                    pypi_0    pypi
[conda] tensorflow                1.15.0          mkl_py37hb249377_0  
[conda] tensorflow-base           1.15.0          mkl_py37h032239d_0  
[conda] torchsummary              1.5.1                    pypi_0    pypi
[conda] torchvision               0.7.0                  py37_cpu    pytorch

## Additional context

[crashreport.txt](https://github.com/pytorch/pytorch/files/5057781/crashreport.txt) generated automatically by MacOs.

cc @suo @gmagogsfm

**Comment by malfet** [comment]
Can not reproduce the crash on Linux nightly

**Comment by hal-314** [comment]
I have the same error (`Segmentation fault (core dumped)` ) with Ubuntu 20.04 and Pytorch 1.5.1 or 1.6. Here is my machine specs:

PyTorch version: 1.5.1
Is debug build: No
CUDA used to build PyTorch: 10.2

OS: Ubuntu 20.04 LTS
GCC version: (Ubuntu 9.3.0-10ubuntu2) 9.3.0
Clang version: Could not collect
CMake version: Could not collect

Python version: 3.7 (64-bit runtime)
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce RTX 2080 Ti
Nvidia driver version: 440.100
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.18.5
[pip3] torch==1.5.1
[pip3] torchvision==0.6.0a0+35d732a
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.2.89              hfd86e86_1  
[conda] mkl                       2020.1                      217  
[conda] mkl-service               2.3.0            py37he904b0f_0  
[conda] mkl_fft                   1.1.0            py37h23d657b_0  
[conda] mkl_random                1.1.1            py37h0573a6f_0  
[conda] numpy                     1.18.5           py37ha1c710e_0  
[conda] numpy-base                1.18.5           py37hde5b4d6_0  
[conda] pytorch                   1.5.1           py3.7_cuda10.2.89_cudnn7.6.5_0    pytorch
[conda] torchvision               0.6.1                py37_cu102    pytorch

Note: I tested Pytorch 1.6 with a new conda environment with only pytorch (`conda install pytorch torchvision cudatoolkit=10.2 -c pytorch`).

**Comment by SplitInfinity** [comment]
I think #41509 added code that prevents a segfault in this case but the error message it shows still not helpful. However, it did not make it into 1.6, and that's probably why @hal-314 is seeing a segfault while @malfet is not. I fixed a similar issue in #38130 where there was a segfault if the type annotation was `List`, so I'll take a look at this and post a PR later today.
